src/persist/src/indexed/future.rs

```diff
@@ -0,0 +1,170 @@
+//! updates, indexed by time.
+
+use std::sync::Arc;
+
+use timely::progress::Antichain;
+use timely::progress::Timestamp;
+
+use crate::error::Error;
+use crate::indexed::cache::BlobCache;
+use crate::indexed::BlobFutureBatch;
+use crate::indexed::BlobFutureMeta;
+use crate::persister::Snapshot;
+use crate::storage::Blob;
+use crate::storage::SeqNo;
+
+/// A persistent, compacting data structure containing `(Key, Value, Time,
+/// Diff)` entries indexed by `(time, key, value)`.
+///
+/// Invariants:
+/// - All entries are after some time frontier and less than some SeqNo.
+///   Intuitively, this is the data that has been drained out of a
+///   [crate::storage::Buffer] but not yet "seal"ed into a
+///   [crate::indexed::trace::BlobTrace].
+/// - TODO: Space usage.
+pub struct BlobFuture {
+    frontier: Antichain<u64>,
+    // NB: The SeqNo ranges here are sorted and contiguous.
+    batches: Vec<((SeqNo, SeqNo), String)>,
+}
+
+impl Default for BlobFuture {
+    fn default() -> Self {
+        BlobFuture::new(BlobFutureMeta {
+            frontier: Antichain::from_elem(Timestamp::minimum()),
+            batches: Vec::new(),
+        })
+    }
+}
+
+impl BlobFuture {
+    /// Returns a BlobFuture re-instantiated with the previously serialized
+    /// state.
+    pub fn new(meta: BlobFutureMeta) -> Self {
+        BlobFuture {
+            frontier: meta.frontier,
+            batches: meta.batches,
+        }
+    }
+
+    /// Serializes the state of this BlobFuture for later re-instantiation.
+    pub fn meta(&self) -> BlobFutureMeta {
+        BlobFutureMeta {
+            frontier: self.frontier.clone(),
+            batches: self.batches.clone(),
+        }
+    }
+
+    /// Returns the smallest SeqNo greater than all the ones contained.
+    pub fn seqno(&self) -> SeqNo {
+        self.batches.last().map_or(SeqNo(0), |((_, end), _)| *end)
+    }
+
+    /// Writes the given batch to [Blob] storage at the given key and logically
+    /// adds the contained updates to this future.
+    pub fn append<L: Blob>(
+        &mut self,
+        key: String,
+        batch: BlobFutureBatch,
+        blob: &mut BlobCache<L>,
+    ) -> Result<(), Error> {
+        if batch.desc.0 != self.seqno() {
+            return Err(Error::from(format!(
+                "batch lower doesn't match seqno {:?}: {:?}",
+                self.seqno(),
+                batch.desc
+            )));
+        }
+        // TODO: Sort the updates in the batch by `(time, key, value)` (or
+        // ensure that they're sorted, if it turns out this work should have
+        // happened somewhere else).
+        let desc = batch.desc.clone();
+        blob.set_future_batch(key.clone(), batch)?;
+        self.batches.push((desc, key));
+        Ok(())
+    }
+
+    /// Returns a consistent read of the updates contained in this future
+    /// matching the given filters (in practice, everything not in Buffer or
```

Comment:
should this be "everything not in Trace"? as in -- if an update is in Trace and Future, we prefer reading from Trace. Similarly, if an update is in Future and Buffer we prefer reading from Future?

Reply:
Yeah, you're totally right. That's better.

Huh, apparently this is even how I documented it in Indexed: https://github.com/MaterializeInc/materialize/pull/6906/files#diff-e5b59bb186539629a1557d1587cdb800a190bf75842900abab2b2d33018926b2R56-R58